### DISCUSSION TITLE
Training and fine-tuning YourTTS model

### DISCUSSION THREAD

**Discussion by annaklyueva** [coqui-ai/TTS#1249] [General Q&A]
Hello!

Aproximately 1 month ago you mentioned that soon there will be a recipe to fine-tune and train YourTTS model. 

Today I was searching but wasn't able to find it. Could you provide a link to it please? Thank you!

I'm asking because I wanted to train this model for a new language but have problems with it.

All the best, 
Anna

**Comment by erogol** [comment]
There is already a recipe under recipes/multilingual

**Reply by e0xextazy** [reply]
Could you give a brief explanation of what's going on there? For example from which checkpoint the training resumes, on which data and things like that

**Reply by hemath1001** [reply]
Sorry but... [here](https://github.com/coqui-ai/TTS/discussions/1058#discussioncomment-1925808) WeberJulian replied that "The recipe under multilingual/vits isn't the one used for YourTTS". I'm confused. (I know YourTTS is based on VITS, so the recipe is? Maybe... another way to train but not exactly YourTTS, I guess?)

**Reply by erogol** [reply]
we don't plan to release the exact recipe until we find a solution to fight against the malicious use of the model. 

Currently we want to train a synthetic voice detector before we release the exact recipe.

**Reply by OmiWakode** [reply]
Hi @erogol . I would like to thank you for the amazing work! I am new to TTS domain, and I would like to know when are you planning to release the exact recipe for the YourTTS model? I want to train a custom YourTTS model on my data set.  Would love to hear back

**Reply by erogol** [reply]
no time schedule yet.

**Reply by twobob** [reply]
You wont release the recipe until you fix the chance of a malicious actor doing bad with it.  So, never then. Fair enough.

**Reply by Kaoru8** [reply]
> You wont release the recipe until you fix the chance of a malicious actor doing bad with it. So, never then. Fair enough.

I know, right? Calling it "YourTTS" and then not allowing anyone to use it is about as ridiculous as "OpenAI" being called that when they deliberately sabotage and restrict access to their models for the same flawed reasoning.

What's the point of even publishing a paper, mentioning it in the documentation, and boasting about how SOTA it is... when it functionally does not exist? Do or do not, this limbo over 0.01% of potential users abusing it is ridiculous. Some people will use a kitchen knife to kill someone, you don't see knife manufacturers ceasing production until foolproof anti-stabbing clothing is developed first.

**Reply by FrontierDK** [reply]
I second this. All TTS products aren't perfect, and I'm guessing a lot of TTS "detectors" already exist, which prohibits abuse. Heck, you can even create a free web service with an AI behind it, allowing anyone to upload/test for any file of TTS traces. Just visited eurodns.com - detect.ai and reveal.ai are already taken.

So not uploading/sharing any recipe won't prohibit anything. At most, it will just delay it for a short while.

**Reply by Kaoru8** [reply]
So, I posted my previous comment right after finally getting YourTTS training to work - after a lot of experimentation, googling that lead nowhere, and other forms of jumping through hoops. It was frustrating as hell, but it was finally working, so I just fired off that quick comment and moved on, thinking I wasted more than enough time on this already and there are other things I have to deal with. But I still can't get it out of my head, because I keep thinking about all the other people what are dealing with the same thing right now, and either entirely waste their time before eventually giving up, or ultimately succeed but only after wasting large amounts of time in the process for absolutely no good reason...

So, first things first... the obfuscation and "security through obscurity" approach is clearly ineffective - I got YourTTS training to work. I wasn't even particularly motivated - I just wanted the convenience of training and running inference for multiple languages on a single model, and the more effective method of adding and training on new speakers was appealing too, since I like constantly experimenting with new voices and the amount of data is always limited. So if I managed to get it working, despite my relatively low motivation and zero previous knowledge of TTS models, you can have a fairly high degree of confidence that anyone intent on abusing the technology could easily do the same, and would be way more motivated to do so - especially if the type of abuse in any way involved making money. So we can conclude that this approach is utterly ineffective in its stated purpose of preventing abuse, and is therefore pointless. Even if we close our eyes for a moment and pretend it's effective, the people intent on abusing TTS will simply use a different model, whether it's an existing one or and upcoming one. You're not stopping, mitigating, or slowing down anything - you're just moving it elsewhere.

Since you're already aware of that, you pivoted to creating a synthetic voice detection model - a goal that's in direct opposition and conflict with the entirety of TTS. On one hand, you're trying to develop ever-improving TTS models capable of producing speech indistinguishable from real human voices... but on the other, you're developing technology to make it distinguishiable. So sooner or later, one or the other has to break. Are you going to deliberately cripple your TTS models and make them underperform so they can remain detectable, or will the detection model inevitably become useless as you keep improving the TTS? Either way, you're pursuing two mutually opposing goals, one of which will always gradually render the other obsolete and is therefore a waste of time. You're effectively splitting all your time and resources in half, in order to wage a pointless war, on yourself.

As researches and developers, you aren't responsible (criminally or ethically) for any potential abuse of the resulting technology, despite what mass media might tell you in their love of sensationalism and witch hunts. Your role is to research and develop, and ideally share with the rest of the world, the overwhelming majority of which will use it for either benevolent or harmless purposes. Any potential abuse from the remaining and inevitable minority is something for laws, courts and cops to deal with - that's literally why they exist.

In the meantime, with this Schrodinger's model, I'd argue that the ethical implications of your current approach to this are far worse and impact way more people than any potential abuse would. People see YourTTS mentioned everywhere in the documentation, spend time reading a research paper extolling its virtues and advertising how "your" it is... and then they can't find any documentation (official or otherwise) on how to actually train it, spend hours searching Google and GitHub issues trying to troubleshoot it, downloading and testing different project versions, forks, pretrained model versions... only to eventually find an obscure thread like this where a dev basically says "yeah, we just trolled you and thousands of other people and wasted hours of your life by advertising something you're actually not allowed to use, because there's a 0.01% chance that someone somewhere will abuse it in some way at some future point in time. Our measures for preventing that are 100% ineffective, but they apply to you and punish you for no good reason anyway. Sucks to be you."

Look... ultimately, you're the developers, and it's obviously perfectly within your rights to do whatever you want with the project. Burn it all to the ground, if you want. But for the love of God or whatever omnipotent AI deity you believe in, just be upfront about it so people don't waste hours of their life they'll never get back on chasing shadows. If it's officialy unsupported, there should obviously be zero mention of it in the official documentation, or on any other official page/blog/repository/whatever. It either exists and you're making it available, or it doesn't and you're not - don't send people on wild goose chases for absolutely no reason. Again, I could very easily argue that the countless real hours spent by thousands of very real people trying to figure out how to train the illusive and widely advertised YourTTS model, FAR outweigh any entirely hypothetical abuse by a handful of people.

Your intentions are obviously admirable and I'll fundamentally always support those, but your method of attempting to achieve them is utterly flawed. It punishes the vast majority of people for the hypothetical abuse by a few, it's utterly ineffective in its stated purpose of preventing or even mitigating said abuse, and it's all over an issue that isn't yours to deal with in the first place - it's what laws and law-related organizations like courts are for. If you insist on maintaining this approach, that's obviously your right, but then you also have to accept that you have an ethical responsiblity to not waste the time of countless legitimate users by advertising the very thing you're not allowing them to use.

And if you are maintaining your approach, what exactly is your policy on a third party sharing training instructions? You don't want to do it, and I imagine someone else would have already done it, but there was an utter lack of information when I was looking for it, so... If I hypothetically wrote an article on Medium or something with step-by-step instructions on how to prepare datasets and train YourTTS, should I expect a cease-and-desist e-mail? A call from a lawyer? A black SUV with dark windows and armed men parked outside my house? :D And if you wouldn't have an issue with it... that would just render your current approach even more pointless, no?

But seriously, if you're refusing to budge on the issue, at least rename the project. "YourTTS" is literally just a cruel joke on any poor bastard that had the misfortune of looking into it after reading the carrot-on-a-stick documentation or paper. And don't even get me started on the whole "TTS for everyone" tagline...

**Reply by alifarrokh** [reply]
Hi @Kaoru8. Can you please share the steps you took for training YourTTS?

**Reply by Kaoru8** [reply]
> Hi @Kaoru8. Can you please share the steps you took for training YourTTS?

Not sure how long it will survive here given the authors' policy regarding this, but if it's just long enough for you to read it and save a few (dozen) hours of your life, it's worth it. Also, disclaimer first - when I say "I got it to work", what I mean precisely is "the pre-training config and model summary looked valid and the training loop properly started". I have yet to properly test it by fully training and evaluating a YourTTS model, because as I said, I had already wasted too much time on it and had to move on (ie. dedicate GPU time) to another major project for the time being. So, what finally got the training loop to run for me:

1) Use the official library, ie. the one you'd get by doing "pip install TTS", no forks/different branches, official or otherwise. Version 0.7.1 works for me. If you're currently using anything else, either purge it all, or create a new virtual environment for this.

2) Download the pretrained "tts_models/multilingual/multi-dataset/your_tts" model + speaker encoder ZIP. Download URL is in "TTS/.models.json". The version that worked for me is here (https://coqui.gateway.scarf.sh/v0.6.1_models/tts_models--multilingual--multi-dataset--your_tts.zip) and references commit e9a1953e as being compatible.

3) Prepare your datasets. For whatever reason, language ID is a dataset-level rather than a file-level setting (like speaker IDs are), so if you have a single dataset with multiple languages, you'll have to split them into separate ones per language first. To be clear, you can have multiple datasets with the same language, just not a single dataset with multiple languages.

4) Copy the config from the ZIP you downloaded, load the copy, and overwrite the following keys: "datasets", "test_sentences", "characters/characters", "characters/punctuations", "model_args/d_vector_file", "model_args/language_ids_file", "model_args/speaker_encoder_config_path", "model_args/speaker_encoder_model_path" to reflect your datasets/environment. Use the downloaded model config as reference. "d_vector_file" doesn't matter at the moment as you'll be generating it in the next step. "model_args/num_speakers" and "model_args/num_languages" may or may not be required, but just set them for good measure.

5) Run "TTS/bin/compute_embeddings.py" using the downloaded speaker encoder model + config (the "_se" files in the ZIP), and your modified training config. The resulting JSON file with speaker embeddings is what your "model_args/d_vector_file" should point to, so edit that into your config.

6) Start training like any other model, resuming from the downloaded pretrained model.

Pretty trivial when it's simply laid out like this, what really killed me was finding a valid combination of repo/commit/fork + pretrained model/speaker encoder/config. Oh, and in the end I had to patch the "PreEmphasis" class in "encoders/model/base_encoder.py" - wrap "return torch.nn.functional.conv1d(x, self.filter).squeeze(1)" in a try/except block, on exception it should "return torch.nn.functional.conv1d(**x.cpu()**, self.filter).squeeze(1)" instead.

**Reply by alifarrokh** [reply]
Can you explain the format of the dataset too? This [official documentation](https://tts.readthedocs.io/en/latest/formatting_your_dataset.html) doesn't provide any information about multi-speaker datasets.

Thanks in advance

@Kaoru8

**Reply by Kaoru8** [reply]
@alifarrokh There is no set format for datasets. You prepare them however you want, making sure to include all relevant info like filename, text, and in this case speaker names. Then you write a custom formatter that reads it and loads it into a dict formatted in a way the library expects. Info is in the link, and check existing formatters in the source code.

**Reply by Ca-ressemble-a-du-fake** [reply]
@Kaoru8 Thanks for sharing your training method. I tried your patch (and also **x.cuda()** because I need the model to be on GPU) without success. I am still facing "input type (cuda) and weight type should be the same" which means the model is for whatever reason on cpu instead of GPU. Do you have any idea why and how I could force model to be loaded on GPU (if that's the problem) ? I tried to prepend my command with `CUDA_VISIBLE_DEVICE=0" without success either.

**Reply by tamerxkilinc** [reply]
@Kaoru8 How about creating a repository with your training script and all its assets? Would definitely help some people here :)
